### PR TITLE
Add html parsing and web scraping

### PR DESCRIPTION
**Pull request by mattlindsey**
For issue #68.

Still work in progress, but hoping for feedback.  Try importing a page from a site.  Need to fix tests.

### CODE REVIEW

**Comment by nickthecook** [comment]
Thanks for the PR for such a cool feature. I'll take a look this morning!

File: app/controllers/documents_controller.rb
Comment (nickthecook) [inline app/controllers/documents_controller.rb:68]: I had to change `file.path` to `f.path` here to get it to run.

**Review by nickthecook — COMMENTED** [review]
Looking good! This fits really well into the existing parser structure.

I tried it with the Wikipedia article on LLMs, and it worked.

Seems like it will fit into Document well enough too; are you thinking of making `filename` optional, or maybe renaming it to `title` and trying to populate it with the title of the page?

**Comment by nickthecook** [comment]
Yep, that's my fault. I introduced a bug in Setting that keeps creating duplicate settings. I just ran into it myself.

Go to the admin ui, then Settings, then delete any settings with nil values.

I'll work on a fix.

**Comment by nickthecook** [comment]
I think the root of the problem is that I added validation to Setting so it wouldn't create duplicates, but some duplicates already existed in our databases because of a previous bug.

`rails db:seed` will deduplicate them for you now.

**Comment by mattlindsey** [comment]
Great. Works now.
You can merge this PR if you want. Or if you think there's more to be done, I can keep working on it.

**Comment by mattlindsey** [comment]
Actually don't merge this yet.  
The chunking for plain text and markdown seems to be broken. Looks like chunk_size isn't getting set and may be defaulting to 1000, but I don't know why.  Take a look if you can. Thanks.

**Comment by mattlindsey** [comment]
Fixed this.  Everthing seems good now.

**Comment by nickthecook** [comment]
> Fixed this. Everthing seems good now.

Great! I'll have another look, and comment on code organization/UI stuff this time, since it's out of draft.

File: app/controllers/documents_controller.rb
Comment (nickthecook) [inline app/controllers/documents_controller.rb:70]: This should include the exception message in the `error_message` field. The error message will show when the user hovers over the "Error" pill on the document in the UI.

File: app/controllers/documents_controller.rb
Comment (nickthecook) [inline app/controllers/documents_controller.rb:63]: This might be better in a separate method.

File: app/controllers/documents_controller.rb
Comment (nickthecook) [inline app/controllers/documents_controller.rb:68]: The tempfile names aren't very descriptive. What do you think about grabbing the page title and using that as the document filename?

Only the `file` attribute is managed by ActiveStorage; I don't think `file` and `filename` have to match.

File: app/views/documents/_form.html.erb
Comment (nickthecook) [inline app/views/documents/_form.html.erb:9]: This looks good in light mode, but in dark mode the text is light but the bg is white.

Can you theme it like the search bar on the same page?

Comment (nickthecook) [reply app/views/documents/_form.html.erb:9]: Also, it would be good if there was some hint for the user as to what the link field was. It does say "Add document or URL" above the file upload box, but maybe if you added placeholder text it would help?

**Review by nickthecook — COMMENTED** [review]
Looking good! Just a few comments.

File: app/controllers/documents_controller.rb
Comment (mattlindsey) [reply app/controllers/documents_controller.rb:68]: I tried that but it was kinda clunky and a html file on disk can have both a filename and title, so I'd rather add `title` to the documents table and display that instead of filename if it is not blank, if that's ok.

File: app/views/documents/_form.html.erb
Comment (nickthecook) [reply app/views/documents/_form.html.erb:9]: This looks good in dark mode:

```suggestion
          <%= f.url_field "link",
            placeholder: "Enter a URL",
            class: "w-full p-2 rounded-lg my-2 px-5 py-2 border border-secondary-400 dark:bg-secondary-925 dark:border-secondary-600"
          %>
```

**Review by nickthecook — APPROVED** [review]
Thanks, @mattlindsey !

**Comment by nickthecook** [comment]
I think somehow the change you made to set the document title was lost.

Any chance you can find that in a branch and push it again?

**Comment by nickthecook** [comment]
Changes to set `Document#error_message` and move the link stuff to a separate method in `DocumentsController` are also lost.

**Pull request merged**
